controller/src/test/java/io/pravega/controller/store/task/TaskStoreFactoryForTests.java

```diff
@@ -0,0 +1,141 @@
+/**
+ * Copyright (c) 2019 Dell Inc., or its subsidiaries. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ */
+package io.pravega.controller.store.task;
+
+import io.pravega.controller.store.client.StoreClient;
+import io.pravega.controller.task.TaskData;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.NotImplementedException;
+import org.apache.curator.framework.CuratorFramework;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Class that extends the functionality of {@link TaskMetadataStore} for test purposes.
+ */
+@Slf4j
+public class TaskStoreFactoryForTests {
+
+    public static TaskMetadataStore createStore(StoreClient storeClient, ScheduledExecutorService executor) {
+        switch (storeClient.getType()) {
+            case Zookeeper:
+            case PravegaTable:
+                return new ZKTaskMetadataStoreForTests((CuratorFramework) storeClient.getClient(), executor);
+            case InMemory:
+                return new InMemoryTaskMetadataStoreForTests(executor);
+            default:
+                throw new NotImplementedException(storeClient.getType().toString());
+        }
+    }
+
+    public static TaskMetadataStore createInMemoryStore(ScheduledExecutorService executor) {
+        return new InMemoryTaskMetadataStoreForTests(executor);
+    }
+
+    /**
+     * Enables additional functionality for testing in TaskMetadataStore
+     */
+    interface TaskMetadataStoreExtensions {
+        /**
+         * Blocks a critical section
+         */
+        void blockCriticalSection();
+
+        /**
+         * Unblocks a critical section
+         */
+        void unblockCriticalSection();
+    }
+
+    /**
+     * Extends {@link ZKTaskMetadataStore} with the calls in {@link TaskMetadataStoreExtensions}.
+     */
+    public static class ZKTaskMetadataStoreForTests extends ZKTaskMetadataStore implements TaskMetadataStoreExtensions {
+
+        private final AtomicReference<CompletableFuture<Void>> latch = new AtomicReference<>(new CompletableFuture<Void>());
```

Comment:
nit: the referenced future is completed in the constructor and not serving any purpose. We can simply do a null future here.

Reply:
I did it this way so that we can have a complete future that does not hold the processing of the lock in the case the test is not blocking it, which is the case for most tests as of now. If I set it to `null` during initialization, then I'd need to verify whether the value is `null` every time. This latter approach works too, but I felt it performs slightly worse.